sdk/client.py

```diff
@@ -147,20 +43,58 @@ def __init__(
         token_provider: Optional[TokenProvider] = None,
         endpoints: Optional[Endpoints] = None,
     ):
-        if not (username and password) and not token_provider:
-            raise Exception(
-                "At least a username/password combination or"
-                "TokenProvider object should be provided."
+        if not username and not token_provider:
+            raise ValueError(
+                "At least a username or TokenProvider object should be provided."
             )
-        self.endpoints = endpoints or Endpoints()
-        self.group_id = group_id
+        self._check_token_provider(token_provider)
 
-        if username and password:
-            token_provider = UsernamePasswordTokenProvider(
-                username, password, f"{self.endpoints.account}/api/v1/auth/login"
-            )
+        if username:
+            token_provider = self._credential_login(username, password)
 
         self.authenticator = HTTPBearerAuthenticator(token_provider)
+        self.endpoints = self._make_endpoints(endpoints)
+        self.group_id = group_id
+
+    @staticmethod
+    def _make_endpoints(endpoints: Optional[Endpoints]) -> Endpoints:
+        if endpoints is None:
+            return Endpoints()
+
+        if not isinstance(endpoints, Endpoints):
+            raise ValueError("Endpoints must be a Endpoints instance")
+
+        return endpoints
+
+    @staticmethod
+    def _check_token_provider(token_provider: Optional[TokenProvider]) -> None:
+        if not token_provider:
+            return
```

Comment:
nit: Personnal preference but I would not allow None token_provider in the `check_token_provider` function since strictly speaking None is not a valid value, instead I would call this function in the init function only if the user inputs a token_provider.

Reply:
done